lightshow/postprocess/autobroaden.py

```diff
@@ -0,0 +1,289 @@
+        
+        yout = np.zeros_like(x)
+        
+        weight = 0
+            
+        for site_spectrum in site_spectra:
+            xin, yin = site_spectrum.spectrum
+        
+            if self.EXPshift != None:
+                xin = xin + self.EXPshift+site_spectrum.VASP_out[0]-site_spectrum.VASP_out[1]-site_spectrum.VASP_out[2]+self.core_state
+            else:   
+                xin = xin + shift
+                
+            if cross_sec:
+                yin = yin * xin
+                
+            if broad_method == "gauss":
+         
+                yout=yout*weight + self.gauss_broaden(x,xin,yin,sigma)*site_spectrum.weight
+            
+            elif broad_method == "lorentz":
+
+                yout=yout*weight + self.lorentz_broaden(x,xin,yin,gamma)*site_spectrum.weight
+            
+            elif broad_method == "voigt":
+            
+                yout=yout*weight + self.voigt_broaden(x,xin,yin,sigma,gamma)*site_spectrum.weight
+            
+            weight = weight + site_spectrum.weight            
+
+        return yout/weight 
+     
+    
+    
+class Autobroaden():
+    def __init__(
+        self,
+        broaden_paras: [float, float],
+        CHlifetime : float,
+        core_state : float,
+        EXPshift = None
+    ):
+
+        """
+        Args:
+            broaden_paras: Broadening parameters, sigma of Gaussian and lorentz dividor 
+            CHlifetime: Core hole life time, Zn K-edge: 1.67, Ti K-edge: 0.89, 
+              taken from Campbell, J. and T. Papp, Widths of the atomic K–N7 levels. Atomic Data and Nuclear Data Tables, 2001. 77(1): p. 1-56.
+            core_state: Core state, e.g., 1s state energy for K-edge
+            EXPshift: Empirical constant shift to match theory to experiment
+        """
+
+        
+        self.broaden_paras = broaden_paras
+        self.CHlifetime = CHlifetime
+        self.core_state = core_state
+        self.EXPshift = EXPshift   
+
+        
+    @property
+    def sigma(self):
+        return self.broaden_paras[0]
+
+    @property
+    def lorentz_divider(self):
+        return self.broaden_paras[1]
+    
+    
+    @property
+    def empirical_shift(self):
+        return self.EXPshift 
+    
+  
+    @staticmethod
+    def deriv(x,y):
+    
+        yout = []
+
+        for i in range(len(x)-1):
+            yout.append((y[i+1]-y[i])/(x[i+1]-x[i]))
+
+        yout.append(0)
+        #yout = np.array(yout)
+        
+        return yout
+        
+
+    @staticmethod
+    def voigt_broaden(x,xin,yin,sigma,ld,lt,ef):
+        yout = np.zeros_like(x)
+        
+        num=len(xin)
+        for i in range(num):
+            gamma=lt/2
+            if xin[i] > ef:
+                gamma=gamma+(xin[i]-ef)/ld/2
+            yout=yout+yin[i]*voigt(x-xin[i],sigma,gamma)/num   
+        return yout 
+    
+
+    def crystal_spectrum_broaden(self, x, site_spectra, sigma, ld, shift , cross_sec = True):
```

Comment:
Try not to repeat code too much. Maybe define a common function with the code common to both?